ocd-2679.sql

```diff
@@ -0,0 +1,65 @@
+-- Create new table to support relationship between user and certification_body table
+DROP TABLE IF EXISTS openchpl.user_certification_body_map;
+CREATE TABLE IF NOT EXISTS openchpl.user_certification_body_map (
+	id bigserial NOT NULL,
+	user_id bigint NOT NULL,
+	certification_body_id bigint NOT NULL,
+	retired bool NOT NULL DEFAULT false,
+	creation_date timestamp NOT NULL DEFAULT NOW(),
+	last_modified_date timestamp NOT NULL DEFAULT NOW(),
+	last_modified_user bigint NOT NULL,
+	deleted bool NOT NULL DEFAULT false,
+	CONSTRAINT user_certification_body_pk PRIMARY KEY (id),
+	CONSTRAINT user_fk FOREIGN KEY (user_id)
+		REFERENCES openchpl.user (user_id) 
+		MATCH SIMPLE ON UPDATE NO ACTION ON DELETE RESTRICT,
+	CONSTRAINT certification_body_fk FOREIGN KEY (certification_body_id)
+		REFERENCES openchpl.certification_body (certification_body_id) 
+		MATCH SIMPLE ON UPDATE NO ACTION ON DELETE RESTRICT
+);
+
+-- Populate the new table based on the ACL tables
+INSERT INTO openchpl.user_certification_body_map
+(user_id, certification_body_id, last_modified_user)
+select sid.id, oi.object_id_identity, -2
+from openchpl.acl_sid sid
+	inner join openchpl.acl_entry entry
+		on sid.id = entry.sid
+	inner join openchpl.acl_object_identity oi
+		on entry.acl_object_identity = oi.id
+	inner join openchpl.acl_class c
+		on oi.object_id_class = c.id
+where c.class = 'gov.healthit.chpl.dto.CertificationBodyDTO'
+and exists (select * from openchpl.user where user_id = sid.id);
+
+-- Backup acl_object_identity table
+CREATE TABLE openchpl.acl_object_identity_backup AS
+SELECT * FROM openchpl.acl_object_identity;
+
+-- Backup acl_class table
+CREATE TABLE openchpl.acl_class_backup AS
+SELECT * FROM openchpl.acl_class;
+
+-- Backup acl_entry table
+CREATE TABLE openchpl.acl_entry_backup AS
+SELECT * FROM openchpl.acl_entry;
+
+-- Delete rows from acl_entry relacted to ACBs
+DELETE FROM openchpl.acl_entry
```

Comment:
If the script runs multiple times you probably can't drop the backup tables and also delete the ACB data from the live tables or we'll lose the data. This seems a candidate for our two-part process so this release would drop/create/populate the backup tables and then the next release would delete the now unused ACB data from the spring tables. There is some pre-loaded ACL data you will also want to remove in that following release.

Reply:
I think part of the idea was to remove the data from the ACL tables to make sure there wasn't anything left over from the spring stuff that was still using it, and that's why the backup tables are being created. That said, the multiple run thing is definitely an issue. Maybe this feature will need to be a three step, where the first one doesn't delete the data from the ACL table, but just copies it to the backup, the second deletes the data, and the third deletes the backup tables